rotkehlchen/balances/historical.py

```diff
@@ -61,29 +84,34 @@ def get_asset_balance(self, asset: Asset, timestamp: Timestamp) -> HistoricalBal
 
         May raise:
         - NotFoundError if balance for the asset at the given timestamp does not exist.
+        - DeserializationError if there is a problem deserializing an event or trade from DB.
         """
         events, main_currency = self._get_events_and_currency(to_ts=timestamp, assets=(asset,))
         if len(events) == 0:
             raise NotFoundError(f'No historical data found for {asset} until the given timestamp.')
 
-        amounts = self._calculate_balance_from_events(events)
+        current_balances: dict[Asset, FVal] = defaultdict(FVal)
+        for event in events:
+            if self._update_balances(event=event, current_balances=current_balances) is not None:
+                break
+
         try:
-            price = PriceHistorian().query_historical_price(
+            price = PriceHistorian.query_historical_price(
                 from_asset=asset,
                 to_asset=main_currency,
                 timestamp=timestamp,
             )
         except (RemoteError, NoPriceForGivenTimestamp):
             price = ZERO_PRICE
 
-        return {'amount': amounts[asset], 'price': price}
+        return {'amount': current_balances[asset], 'price': price}
 
     def get_assets_amounts(
             self,
             assets: tuple[Asset, ...],
             from_ts: Timestamp,
             to_ts: Timestamp,
-    ) -> tuple[dict[Timestamp, FVal], str | None]:
+    ) -> tuple[dict[Timestamp, FVal], tuple[str | int | None, str | None] | None]:
```

Comment:
You have to adjust the docstring explaining what the return value is now. As you changed it but the docstring is not updated.